cloudigrade/util/aws/ec2.py

```diff
@@ -105,7 +103,68 @@ def get_ami(session, image_id, source_region):
         Image: A boto3 EC2 Image object.
 
     """
-    return session.resource('ec2', region_name=source_region).Image(image_id)
+    image = session.resource('ec2', region_name=source_region).Image(image_id)
+    check_image_state(image)
+    return image
+
+
+def check_image_state(image):
+    """Raise an exception if image state is not available."""
+    if image.state == 'available':
+        return
+    message = _('Image {id} has state {state} (reason: {reason})').format(
+        id=image.id,
+        state=image.state,
+        reason=image.state_reason
+    )
+    if image.state == 'failed':
+        raise AwsImageError(message)
+    raise ImageNotReadyException(message)
+
+
+def copy_ami(session, image_id, source_region):
+    """
+    Copy an Amazon Machine Image within a given customer account session.
+
+    Args:
+        session (boto3.Session): A temporary session tied to a customer account
+        image_id (str): An AMI ID
+        source_region (str): The region the snapshot resides in
+
+    Returns:
+        str: The image id of the newly created image.
+
+    """
+    old_image = get_ami(session, image_id, source_region)
+    # Note: AWS image names are limited to 128 characters in length.
+    new_name = f'cloudigrade reference copy ({old_image.name})'[:128]
```

Comment:
Not a big deal at all because I actually prefer to format strings this way, but do we want to move this to `' '.format()` to keep it more consistent with the string format we are moving towards?

Reply:
Good point. I will change this to be more consistent.